.github/workflows/chart_push.yml

```diff
@@ -1,23 +1,47 @@
-name: Push Helm Chart to GAR
-on: workflow_dispatch
+name: Publish chart to registry
+on: 
+  push:
+    branches:
+      - 'main'
+    paths:
+      - 'basic-web-service/Chart.yaml'
+  workflow_dispatch: {}
 jobs:
-  google-auth:
+  publish_chart:
     runs-on: ubuntu-latest
+    env:
+      REGISTRY_HOSTNAME: "us-central1-docker.pkg.dev"
     steps:
     - uses: 'actions/checkout@v3'
+    - name: Get latest chart version number
+      run: echo "CHART_VERSION=$(cat basic-web-service/Chart.yaml | grep 'version:' | cut -d' ' -f2)" >> $GITHUB_ENV
+    - name: Check for matching archive
+      env:
+        ERROR_MISSING_ARCHIVE: "::error:: Couldn't find the file 'charts/basic-web-service-${{ env.CHART_VERSION }}.tgz', did you run './package-chart.sh'?"
+      run: |
+        if [[ ! -f  "charts/basic-web-service-${CHART_VERSION}.tgz" ]]
+        then
+          echo $ERROR_MISSING_ARCHIVE && exit 1
+        fi
     - uses: azure/setup-helm@v3
-    - id: 'auth'
-      name: 'Authenticate to Google Cloud'
-      uses: 'google-github-actions/auth@v1'
       with:
-        credentials_json: '${{ secrets.GAR_TOKEN_TEST }}'
-    # Install gcloud, `setup-gcloud` automatically picks up authentication from `auth`.
-    - name: 'Set up Cloud SDK'
-      uses: 'google-github-actions/setup-gcloud@v1'
-    - name: 'Test Helm login'
+        version: 'latest'
+        token: ${{ secrets.GITHUB_TOKEN }} # only needed if version is 'latest'
+    - name: 'Authenticate to Google Artifact Registry'
+      env:
+        GAR_KEY_BASE64: ${{ secrets.GAR_KEY_BASE64 }}
+        REGISTRY_URL_HTTP: "https://${{ env.REGISTRY_HOSTNAME }}"
+      run: |
+        echo $GAR_KEY_BASE64 | helm registry login -u _json_key_base64 --password-stdin \
+        $REGISTRY_URL_HTTP
+    - name: 'Check if chart version already exists in registry'
+      env:
+        CHART_URL_OCI: "oci://${{ env.REGISTRY_HOSTNAME }}/uwit-mci-iam/helm-charts-test/basic-web-service"
+        ERROR_CHART_EXISTS: "::error:: basic-web-service:${{ env.CHART_VERSION }} already exists in registry, did you forget to bump the version in Chart.yaml?"
       run: |
-        gcloud auth application-default print-access-token | helm registry login -u oauth2accesstoken \
-        --password-stdin https://us-central1-docker.pkg.dev
+        helm pull $CHART_URL_OCI --version $CHART_VERSION && echo $ERROR_CHART_EXISTS && exit 1 || echo 'Uploading new chart'
```

Comment:
I'm always really bad at these sorts of statements in bash and encourage wrapping the compound statements in `(` `)` symbols for clarity.